[DataPipe] Fixing pickling issues with fork and demux

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67930

Fixes #67848

Differential Revision: [D32222184](https://our.internmc.facebook.com/intern/diff/D32222184)

cc @VitalyFedyunin @ejguan @NivekT

NivekT committed:
[DataPipe] Fixing pickling issues with fork and demux

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fixing pickling issues with fork and demux"


Fixes #67848

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32222184).

ejguan commented on test/test_datapipe.py:
```diff
@@ -862,6 +864,11 @@ def test_fork_datapipe(self):
         self.assertEqual(len(input_dp), len(dp3))
 
+        # Pickle Test:
+        dp1, dp2, dp3 = input_dp.fork(num_instances=3)
+        from torch.utils.data.graph import traverse
+        traverse(dp1)  # This should not raise any error
```
Could we add another test that we iterates `dp1`, `dp2` and `dp3` and then traverse?

NivekT replied:
Done

ejguan reviewed (approved):
LGTM, thanks for fixing the issue.
I only have one comment on test case.

NivekT committed:
Update on "[DataPipe] Fixing pickling issues with fork and demux"


Fixes #67848

Differential Revision: [D32222184](https://our.internmc.facebook.com/intern/diff/D32222184)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32222184).

NivekT committed:
Update on "[DataPipe] Fixing pickling issues with fork and demux"


Fixes #67848

Differential Revision: [D32222184](https://our.internmc.facebook.com/intern/diff/D32222184)

cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fixing pickling issues with fork and demux"


Fixes #67848

Differential Revision: [D32222184](https://our.internmc.facebook.com/intern/diff/D32222184)

cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32222184).

ejguan commented on test/test_datapipe.py:
```diff
@@ -52,6 +52,7 @@
     runtime_validation,
     runtime_validation_disabled,
 )
+from torch.utils.data.graph import traverse
```
Side note: Do you think we can also add this to one of the standard tests for each DataPipe in the future?
Maybe open an issue?

NivekT replied:
I think it should go here:
https://github.com/pytorch/pytorch/wiki/DataPipes-testing-requirements

One of these two?
- Test if DataPipes creates properly linked DataPipes graph
- Test if DataPipe is picklable (using pickle or dill)

ejguan replied:
I mean we need to test all existing and in-development DataPipes for this case.

ejguan replied:
After iteration, it's still picklable.

NivekT replied:
I see, I will add it to the testing requirement page. 

Perhaps we should have an issue to ensure all DataPipes have tests that meet all the requirements on that page? Or do you prefer to have a specific issue about being picklable after iteration for now?
I can do either.

ejguan replied:
I would go with former one and categorize which test is missing for each DataPipe
